openlibrary/plugins/admin/code.py

```diff
@@ -249,6 +251,69 @@ def GET(self):
         pass
 
 
+class unlink_ia_ol(delegate.page):
+    path = '/ia/unlink'
+    encoding = 'json'
+
+    def POST(self):
+        i = web.input(digest="", msg="")
+
+        digest = i.digest
+        msg = i.msg
+
+        try:
+            HMACToken.verify(digest, msg, "ia_sync_secret")
+        except (ValueError, ExpiredTokenError):
+            raise web.HTTPError("401 Unauthorized")
+
+        ocaid, ts = msg.split("|")
+
+        if not ts or not ocaid:
+            raise web.HTTPError(
+                "400 Bad Request", data=json.dumps({"error": "Invalid inputs"})
+            )
+
+        # Fetch affected editions
+        if not (
+            edition_keys := web.ctx.site.things(
+                {"type": '/type/edition', "ocaid": ocaid}
+            )
+        ):
+            raise web.HTTPError("404 Not Found")
+
+        editions = [web.ctx.site.get(key) for key in edition_keys]
+        if len(editions) > 1:
+            raise web.HTTPError(
+                "409 Conflict",
+                data=json.dumps(
+                    {"error": "Multiple editions associated with given ocaid"}
+                ),
+            )
+
+        edition = editions[0]
+
+        # Update records
+        try:
+            self.make_dark(edition)
+        except ClientException as e:
+            raise web.HTTPError(
+                "500 Internal Server Error", data=json.dumps({"error": str(e)})
+            )
+
+        return delegate.RawText(json.dumps({"status": "ok"}))
+
+    def make_dark(self, edition):
+        data = edition.dict()
+        del data["ocaid"]
+        source_records = data.get("source_records", [])
+        data['source_records'] = [
+            rec for rec in source_records if not rec.startswith("ia:")
+        ]
+        if not data['source_records']:
+            del data['source_records']
```

Comment:
If the `source_records` list is empty after the MARC record is unlinked, `source_records` is deleted from the edition object.